gobblin-core/src/main/java/gobblin/source/extractor/hadoop/CopySource.java

```diff
@@ -0,0 +1,206 @@
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+
+/**
+ * {@link gobblin.source.Source} that generates work units from {@link gobblin.data.management.copy.CopyableDataset}s.
+ * Used for doing byte-level copies of files.
+ */
+public class CopySource extends AbstractSource<String, byte[]> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CopySource.class);
+
+  private static final String CONFIG_PREFIX = "gobblin.copy.source.";
+  static final String SERIALIZED_COPYABLE_FILE = CONFIG_PREFIX + "serialized.copyable.file";
+  static final String COPYABLE_FILE_PATH = CONFIG_PREFIX + "copyable.file.path";
+
+  /**
+   * Parses origin path for a work unit representing a {@link gobblin.data.management.copy.CopyableFile}.
+   * @param state work unit state.
+   * @return path of the file that should be copied.
+   */
+  public static Path getOriginPath(WorkUnitState state) {
+    return new Path(state.getProp(COPYABLE_FILE_PATH));
+  }
+
+  /**
+   * Parses a {@link gobblin.data.management.copy.CopyableFile} from a work unit containing a serialized
+   * {@link gobblin.data.management.copy.CopyableFile}.
+   * @param state work unit state.
+   * @return Deserialized {@link gobblin.data.management.copy.CopyableFile}.
+   * @throws IOException
+   */
+  public static CopyableFile getCopyableFile(WorkUnitState state) throws IOException {
+    return deserializeCopyableFile(state.getProp(SERIALIZED_COPYABLE_FILE));
+  }
+
+  /**
+   * Does the following:
+   *  1. Instantiate a {@link DatasetFinder}.
+   *  2. Find all {@link Dataset} using {@link DatasetFinder}.
+   *  3. For each {@link CopyableDataset} get all {@link CopyableFile}s.
+   *  4. Create a {@link WorkUnit} per {@link CopyableFile}.
+   * @param state see {@link gobblin.configuration.SourceState}
+   * @return Work units for copying files.
+   */
+  @Override public List<WorkUnit> getWorkunits(SourceState state) {
+
+    List<WorkUnit> workUnits = Lists.newArrayList();
+
+    try {
+      FileSystem originFs = getSourceFileSystem(state);
+      FileSystem targetFs = getTargetFileSystem(state);
+
+      DatasetFinder datasetFinder = DatasetUtils.instantiateDatasetFinder(state.getProperties(), originFs);
+      Collection<Dataset> datasets = datasetFinder.findDatasets();
+
+      for (Dataset dataset : datasets) {
+
+        Collection<Partition<CopyableFile>> partitions;
+        if(dataset instanceof CopyableDataset && dataset instanceof PartitionableDataset &&
```

Comment:
Can we think of a way to avoid these casts?

- Since we know that this is a CopySource, may be we should use the CopyDatasetFinder instead of DatasetFinder which vends a bunch of CopyableDatasets and avoid cast from Dataset to CopyableDataset. Later implementations can extends CopyableDatasetFinder.

- Also in the current implementation if the dataset is not partitionable, we create a single partition with all the files. Can the CopyableDataset always extend PartitionableDataset? A copyableDataset will return Partitions instead of files. In case there are no partitions, the CopyableDataset will return a single partition will all files. This will avoid a cast to PartitionableDataset.